src/lasr_labs_2025_control_project/numerical_pipeline/utils.py

```
@@ -0,0 +1,676 @@
+
+        # Record history
+        history.append(
+            {
+                "iteration": iteration + 1,
+                "objective": float(current_value),
+                "params": {k: float(v) for k, v in current_params.items()},
+            }
+        )
+
+        # Check for convergence
+        if abs(current_value - previous_value) < tolerance:
+            break
+
+    return dict(
+        best_score=float(current_value),
+        best_parameters={k: float(v) for k, v in current_params.items()},
+        history=history,
+    )
+
+
+def calculate_tpr_fpr_from_scores(
+    positive_scores: List[float],
+    negative_scores: List[float],
+    num_thresholds: int = 100,
+) -> Tuple[List[float], List[float]]:
+    """Calculate TPR and FPR values from positive and negative scores.
+
+    Args:
+        positive_scores: Scores for positive samples (ground truth = 1)
+        negative_scores: Scores for negative samples (ground truth = 0)
+        num_thresholds: Number of threshold points to evaluate
+
+    Returns:
+        Tuple of (tpr_values, fpr_values) lists
+    """
+    all_scores = np.array(positive_scores + negative_scores)
+    labels = np.concatenate(
+        [np.ones(len(positive_scores)), np.zeros(len(negative_scores))]
+    )
+
+    # Create thresholds from min to max score
+    min_score = np.min(all_scores)
+    max_score = np.max(all_scores)
+    thresholds = np.linspace(min_score, max_score, num_thresholds)
+
+    tpr_values = []
+    fpr_values = []
+
+    for threshold in thresholds:
+        predictions = (all_scores >= threshold).astype(int)
+
+        # Calculate confusion matrix components
+        tp = np.sum((predictions == 1) & (labels == 1))
+        fp = np.sum((predictions == 1) & (labels == 0))
+        tn = np.sum((predictions == 0) & (labels == 0))
+        fn = np.sum((predictions == 0) & (labels == 1))
+
+        # Calculate TPR and FPR
+        tpr = tp / (tp + fn) if (tp + fn) > 0 else 0.0
```

If the else statement gets hit here, should we just throw an error?